### ISSUE TITLE
Is there any particular reason why you use softplus to make sure the scale is non-negative?

### ISSUE THREAD

**Issue by nbro** [tensorflow/probability#751]
In the article https://blog.tensorflow.org/2019/03/regression-with-probabilistic-layers-in.html, you use the softplus to make sure the scales are non-negative. However, if the inputs are e.g. -40.0, the softplus returns values which are extremely small

```
import tensorflow as tf
tf.math.softplus(-40.0)
Out[9]: <tf.Tensor: id=11, shape=(), dtype=float32, numpy=4.248354e-18>
```

which would make the scale extremely small and quite unrealistic (and this can easily lead to NaNs or vanishing and exploding gradients).

Is there any particular reason why you use softplus to make sure the scale is non-negative? Is there any formal study that states that the softplus is an appropriate option in this case? 

For example, why not simply using the absolute value, the ReLU or any other function that doesn't behave badly in the case of negative numbers (like softplus) to convert the negative numbers to positive ones before feeding them into the scale parameter of a distribution?

**Comment by davmre** [comment]
A couple of nice things about softplus (and `exp`, the other transformation widely used to enforce positivity) are that it's monotonic and everywhere differentiable. You certainly could use `abs`, for example, to enforce positivity, but you'd get weirdness around zero because optimizer steps could end up moving in the opposite of the intended direction. For example, if `x=1` and dy/dx=5---i.e., the gradient wants us to make x smaller to minimize y---then a gradient descent step `x' = x - 5` will overshoot and leave us with `x' = -4` which of course has larger absolute value than we started with. Whether this matters in practice depends on your optimizer, step size, loss surface, etc., but it certainly *can* lead to nonconvergent optimizations. Using a monotonic function like softplus avoids the whole issue.

Meanwhile, something like ReLU is monotonic, but has zero gradient for negative inputs, so if the optimizer ever happens to try a negative input it may get stuck there forever. This might be less of an issue in neural layers because the activation is a function of a stochastic input (as opposed to if we were just optimizing a scale variable directly), but even there 'dead ReLU' nodes do still occur, which is why variants like LeakyRuLU and ELU exist. Note that softplus itself is essentially a smoothed ReLU: my understanding is that the main reason it's not commonly used as an NN activation function is that it involves both an `exp` and a `log`, so it's a bit more expensive to compute.

If infinitesimal scales end up being a problem in practice on a given task, the fix we commonly use is a softplus-and-shift, e.g. `scale = epsilon + tf.math.softplus(unconstrained_scale)`, where `epsilon` is some tiny value like 1e-5 that we are a priori confident is much smaller than the true scale.

**Comment by nbro** [comment]
@davmre But I was getting annoying NaNs because I was using softplus, although I noticed that softplus returns 0 (and not NaN) if its input is extremely small. The problem is that, if you create a normal distribution `d` with a very small scale, `d.log_prob` can easily get extremely small,  large or NaN, so, even though softplus is differentiable, it is probably not the most appropriate function for this task. It's probably just better to clip the inputs to the scale parameter of the distribution, rather than using softplus. Also, in this case, why do you care if softplus is differentiable?

**Comment by davmre** [comment]
> The problem is that, if you create a normal distribution d with a very small scale (returned by softplus), d.log_prob can easily get extremely small, large or NaN

Have you tried adding a small shift to the softplus, as I proposed above?

 > Also, in this case, why do you care if softplus is differentiable?

Thinking through this more, the fact that softplus has nonzero derivative almost everywhere (unlike, say, ReLU) is more important than strict differentiability as such. Nondifferentiability at a finite set of points is usually not a big deal, e.g., LeakyReLU (`lambda x: x if x > 0 else 0.01 * x`) is not differentiable at zero but still has nonzero derivative everywhere else.

**Comment by nbro** [comment]
@davmre I haven't tried it yet, but, if add this epsilon to the softplus, I am afraid that there wouldn't be much difference between softplus and just clip the inputs, for example, in the range `[0.01, 1000]`.

**Comment by nbro** [comment]
@davmre I've just tried it and it seems to solve the NaNs issue. Anyway, if add this epsilon to the softplus, I am afraid that there wouldn't be much difference between softplus and just clip the inputs, for example, in the range `[0.01, 1000]` (which I've also tried, and I've noticed that I don't get anymore the NaNs).

**Comment by davmre** [comment]
Glad things are working! The main difference versus clipping is that clipping has zero gradient for inputs outside of the clipped interval, while a shifted softplus has nonzero gradient for all inputs. Though again, how much this matters in practice is very problem-specific.

**Comment by srvasude** [comment]
@nbro Closing this, since I believe this issue is resolved. As mentioned, a shifted softplus is a good constraint function because it has gradients , monotonic and doesn't grow too fast (compared to something like `exp`). But this can depend on problem, and other constraint functions can be a good choice.
